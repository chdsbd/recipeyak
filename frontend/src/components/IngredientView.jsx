import React from "react"

export default class IngredientView extends React.Component {
  render() {
    const { quantity, name, description, optional } = this.props

    const fmtDescription = description ? ", " + description : ""

    return (
      <p className="listitem-text justify-space-between">
        {quantity} {name}
        {fmtDescription}{" "}
        {optional ? <span className="text-muted">[optional]</span> : ""}
      </p>
    )
  }
}
