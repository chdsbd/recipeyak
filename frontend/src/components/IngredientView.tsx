import React from "react"
import { IIngredient } from "@/store/reducers/recipes"
import { capitalizeUnits } from "@/text"

interface IIngredientVIewProps {
  readonly quantity: IIngredient["quantity"]
  readonly name: IIngredient["name"]
  readonly description: IIngredient["description"]
  readonly optional: IIngredient["optional"]
}

export default function IngredientView({
  quantity,
  name,
  description,
  optional
}: IIngredientVIewProps) {
  const fmtDescription = description ? ", " + description : ""

  return (
    <p className="listitem-text justify-space-between">
      {capitalizeUnits(quantity)} {name}
      {fmtDescription}{" "}
      {optional ? <span className="text-muted">[optional]</span> : ""}
    </p>
  )
}
