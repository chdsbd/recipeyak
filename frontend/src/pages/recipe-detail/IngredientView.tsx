import { DragElementWrapper, DragSourceOptions } from "react-dnd"

import { clx } from "@/classnames"
import { RecipeFetchResponse as Recipe } from "@/queries/recipeFetch"
import { normalizeUnitsFracs } from "@/text"
import {
  THEME_CSS_BAKING_POWDER,
  THEME_CSS_BAKING_SODA,
} from "@/themeConstants"

type Ingredient = Recipe["ingredients"][number]
interface IIngredientVIewProps {
  readonly quantity: Ingredient["quantity"]
  readonly name: Ingredient["name"]
  readonly description: Ingredient["description"]
  readonly optional: Ingredient["optional"]
  readonly dragRef: DragElementWrapper<DragSourceOptions> | undefined
}

export function IngredientViewContent({
  quantity,
  name,
  description,
  optional,
}: Omit<IIngredientVIewProps, "dragRef">) {
  description = description
    ? ", " + normalizeUnitsFracs(description).trim()
    : ""
  name = normalizeUnitsFracs(name.trim())
  quantity = normalizeUnitsFracs(quantity).trim()
  const isBakingSoda = name.toLocaleLowerCase() === "baking soda"
  const isBakingPowder = name.toLocaleLowerCase() === "baking powder"
  return (
    <>
      <span className="font-medium">{quantity}</span>{" "}
      <span
        className={clx(
          isBakingSoda && THEME_CSS_BAKING_SODA,
          isBakingPowder && THEME_CSS_BAKING_POWDER,
        )}
      >
        {name}
      </span>
      {description}{" "}
      {optional ? (
        <span className="text-[--color-text-muted] print:!text-black">
          [optional]
        </span>
      ) : (
        ""
      )}
    </>
  )
}

export default function IngredientView({
  quantity,
  name,
  description,
  optional,
  dragRef,
}: IIngredientVIewProps) {
  return (
    <div
      className="cursor-auto select-text justify-between whitespace-pre-wrap pb-2 leading-5"
      ref={dragRef}
    >
      <IngredientViewContent
        quantity={quantity}
        description={description}
        name={name}
        optional={optional}
      />
    </div>
  )
}
