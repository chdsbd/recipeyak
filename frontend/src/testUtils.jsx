import React from "react"
import { Provider } from "react-redux"
import { DragDropContext } from "react-dnd"
import { MemoryRouter } from "react-router"
import HTML5Backend from "react-dnd-html5-backend"

import { emptyStore as store } from "./store/store"

@DragDropContext(HTML5Backend)
export class DndTestContext extends React.Component {
  render() {
    return (
      <Provider store={store}>
        <MemoryRouter>{this.props.children}</MemoryRouter>
      </Provider>
    )
  }
}
