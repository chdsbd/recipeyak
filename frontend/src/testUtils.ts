import { DefaultBodyType, MockedRequest, rest } from "msw"
import { setupServer } from "msw/node"

import { store } from "@/store/store"

export { rest }

type Deferred<T> = {
  promise: PromiseLike<T>
  done: () => void
  error: () => void
}

// based from: https://stackoverflow.com/a/69027809/3720597
function deferred<T>(data: T): Deferred<T> {
  let resolve: (value: T | PromiseLike<T>) => void
  let reject: (reason?: unknown) => void
  const promise = new Promise<T>((res, rej) => {
    resolve = res
    reject = rej
  })
  return {
    promise,
    done: () => {
      resolve(data)
    },
    error: () => {
      reject(data)
    },
  }
}

beforeEach(() => {
  store.dispatch({ type: "@@RESET" })
})

export const server = setupServer()
// Start server before all tests
beforeAll(() => {
  server.listen()
})
// Close server after all tests
afterAll(() => {
  server.close()
})
// Reset handlers after each test `important for test isolation`
afterEach(() => {
  server.resetHandlers()
})

// We setup all this stuff below to get test failures when we forget to mock an
// endpoint.
// ideally it would be builtin, waiting on: https://github.com/mswjs/msw/issues/946
const inprogress = new Map<string, Deferred<MockedRequest<DefaultBodyType>>>()

// Two options, either a request is mocked or it's not.
// If it's mocked we'll hit the request:match, otherwise we'll hit
// request:unhandled.
server.events.on("request:start", (req) => {
  inprogress.set(req.id, deferred(req))
})
server.events.on("request:match", (req) => {
  inprogress.get(req.id)?.done()
})
server.events.on("request:unhandled", (req) => {
  inprogress.get(req.id)?.error()
})

// check if we had any unhandled requests
afterEach(async () => {
  const errors = (
    await Promise.allSettled([...inprogress.values()].map((x) => x.promise))
  )
    .filter((x): x is PromiseRejectedResult => x.status === "rejected")
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    .map((x): MockedRequest<DefaultBodyType> => x.reason)
  inprogress.clear()
  if (errors.length) {
    // TODO: figure out if we can include where in the code the request was called
    const err = Error(
      errors
        .map(
          (req) =>
            `found an unhandled ${req.method} request to ${req.url.href}`,
        )
        .join("\n\n"),
    )
    // clear the useless stack trace
    err.stack = undefined
    throw err
  }
})
