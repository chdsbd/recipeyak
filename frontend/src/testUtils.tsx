import React from "react"
import { Provider } from "react-redux"
import { DndProvider } from "react-dnd"
import { MemoryRouter } from "react-router"
import HTML5Backend from "react-dnd-html5-backend"

import { emptyStore as store } from "@/store/store"

export class DndTestContext extends React.Component {
  render() {
    return (
      <DndProvider backend={HTML5Backend}>
        <Provider store={store}>
          <MemoryRouter>{this.props.children}</MemoryRouter>
        </Provider>
      </DndProvider>
    )
  }
}
