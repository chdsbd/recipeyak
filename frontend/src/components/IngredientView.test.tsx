import React from "react"
import renderer from "react-test-renderer"
import { TestProvider } from "@/testUtils"
import IngredientView from "@/components/IngredientView"

describe("<IngredientView/>", () => {
  test("snap with capitalization", () => {
    const testCases = [
      "3 tablespoons",
      "1 teaspoon",
      "1 tablespoon + 2 teaspoons",
      "1 Teaspoon"
    ]

    testCases.forEach(quantity => {
      const tree = renderer
        .create(
          <TestProvider>
            <IngredientView
              dragRef={undefined}
              quantity={quantity}
              name="salt"
              optional={false}
              description=""
            />
          </TestProvider>
        )
        .toJSON()

      expect(tree).toMatchSnapshot()
    })
  })
})
