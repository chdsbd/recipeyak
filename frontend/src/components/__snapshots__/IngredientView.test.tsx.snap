// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IngredientView/> snap with capitalization 1`] = `
<p
  className="listitem-text justify-space-between"
>
  3 Tablespoons
   
  salt
  
   
  
</p>
`;

exports[`<IngredientView/> snap with capitalization 2`] = `
<p
  className="listitem-text justify-space-between"
>
  1 teaspoon
   
  salt
  
   
  
</p>
`;

exports[`<IngredientView/> snap with capitalization 3`] = `
<p
  className="listitem-text justify-space-between"
>
  1 Tablespoon + 2 teaspoons
   
  salt
  
   
  
</p>
`;

exports[`<IngredientView/> snap with capitalization 4`] = `
<p
  className="listitem-text justify-space-between"
>
  1 teaspoon
   
  salt
  
   
  
</p>
`;
